File: data_structures/graphs/adjacency_matrix/adjacency_matrix.c
/**
 * @author Yannick Brenning
 * @date 26.12.2021
 * @brief Implementation of graphs using adjacency matrix
 */

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "adjacency_matrix.h"

am_graph_t *am_graph_init(bool directed, int vertices) {
    am_graph_t *graph = (am_graph_t *) malloc(sizeof(am_graph_t));
    assert(graph);

    graph->directed = directed;
    graph->vertices = vertices;
    graph->edges = (bool **) malloc(vertices * sizeof(bool *));
    assert(graph->edges);

    for (int i = 0; i < vertices; i++) {
        graph->edges[i] = (bool *) calloc(vertices, sizeof(bool));
        assert(graph->edges[i]);
    }

    return graph;
}

void am_graph_destroy(am_graph_t **graph) {
    for (int i = 0; i < (*graph)->vertices; i++) {
        free((*graph)->edges[i]);
    }

    free((*graph)->edges);
    free(*graph);
    *graph = NULL;
}

bool am_graph_vertex_exists(am_graph_t *graph, unsigned int vertex) {
    if (vertex < 0 || vertex >= graph->vertices) return false;
    return true;
}

bool am_graph_edge_exists(am_graph_t *graph, unsigned int vertex_1, unsigned int vertex_2) {
    if (graph->directed) return graph->edges[vertex_1][vertex_2];
    else return graph->edges[vertex_1][vertex_2] && graph->edges[vertex_2][vertex_1];
}

bool am_graph_add_edge(am_graph_t *graph, unsigned int vertex_1, unsigned int vertex_2) {
    if (!am_graph_vertex_exists(graph, vertex_1) || !am_graph_vertex_exists(graph, vertex_2)
    || am_graph_edge_exists(graph, vertex_1, vertex_2)) {
        return false;
    }

    graph->edges[vertex_1][vertex_2] = true;
    if (!graph->directed) graph->edges[vertex_2][vertex_1] = true;
    return true;
}

bool am_graph_delete_edge(am_graph_t *graph, unsigned int vertex_1, unsigned int vertex_2) {
    if (!am_graph_vertex_exists(graph, vertex_1) || !am_graph_vertex_exists(graph, vertex_2)
    || !am_graph_edge_exists(graph, vertex_1, vertex_2)) {
        return false;
    }

    graph->edges[vertex_1][vertex_2] = false;
    if (!graph->directed) graph->edges[vertex_2][vertex_1] = false;
    return true;
}

void graph_print(am_graph_t *graph) {
    for (int i = 0; i < graph->vertices; i++) {
        for (int j = 0; j < graph->vertices; j++) {
            if (graph->edges[i][j]) {
                if (graph->directed) printf("%d -> %d\n", i, j);
                else printf("%d -- %d\n", i, j);
            }
        }
    }

    printf("\n");
}

File: data_structures/graphs/adjacency_matrix/adjacency_matrix.py
"""
Implementation of graphs using adjacency matrix.
"""

from __future__ import annotations


class Graph:
    def __init__(self, vertices: int, directed: bool = True) -> None:
        self.directed = directed
        self.vertices = vertices
        self.edges: list[list[bool]] = [[False] * vertices for _ in range(vertices)]

    def __str__(self) -> str:
        output = ""
        for i in range(0, self.vertices):
            for j in range(0, self.vertices):
                if self.edges[i][j]:
                    if self.directed:
                        output += f"{i} -> {j}\n"
                    else:
                        output += f"{i} -- {j}\n"

        return output

    def has_vertex(self, vertex: int) -> bool:
        return 0 <= vertex < self.vertices

    def has_edge(self, vertex_1: int, vertex_2: int) -> bool:
        if not self.has_vertex(vertex_1) or not self.has_vertex(vertex_2):
            return False

        if self.directed:
            return self.edges[vertex_1][vertex_2]

        return self.edges[vertex_1][vertex_2] and self.edges[vertex_2][vertex_1]

    def add_edge(self, vertex_1: int, vertex_2: int) -> bool:
        if self.has_edge(vertex_1, vertex_2):
            return False

        if self.directed:
            self.edges[vertex_1][vertex_2] = True
        else:
            self.edges[vertex_1][vertex_2] = True
            self.edges[vertex_2][vertex_1] = True
        return True

    def delete_edge(self, vertex_1: int, vertex_2: int) -> bool:
        if not self.has_edge(vertex_1, vertex_2):
            return False

        if self.directed:
            self.edges[vertex_1][vertex_2] = False
        else:
            self.edges[vertex_1][vertex_2] = False
            self.edges[vertex_2][vertex_1] = False

        return True


def main():

    # Driver code to generate the following directed graph:
    #
    #           (0) -> (1) -> (2)
    #           |
    #           v
    #          (3)

    graph = Graph(5)
    graph.add_edge(0, 1)
    graph.add_edge(0, 3)
    graph.add_edge(1, 2)
    print(graph)

    graph.delete_edge(0, 1)
    print(graph)


if __name__ == "__main__":
    main()
